core/src/main/java/hudson/model/labels/LabelExpression.java

```diff
@@ -210,4 +225,156 @@ public LabelOperatorPrecedence precedence() {
             return LabelOperatorPrecedence.IMPLIES;
         }
     }
+
+    //region Auto-Completion and Validation
+
+    /**
+     * Utility class for taking the current input value and computing a list of potential terms to match against the
+     * list of defined labels.
+     */
+    static class AutoCompleteSeeder {
```

Comment:
Note that many plugins have copy-pasted the code: https://github.com/search?q=org%3Ajenkinsci+AutoCompleteSeeder&type=Code .
Maybe it;s time to consider making it a public API in `jenkins.model.labels`

Reply:
Makes sense, although it may have been private in part because the auto-completion for labels is still far from perfect - "foo&&b" will pop up completion for the "b" part, but selecting "bar" then replaces the entire expression (mitigated slightly by setting ' ' as delimiter and typing "foo && b").

To me, the fact that many plugins copied the internal implentation just shows how useful exposing an `autoComplete` method is; it does not necessarily make a case for exposing that method's internals.

Having said that, I don't feel all that strongly about _not_ exposing it either.